File: polyVal2D.py
import numpy as np

def polyVal2D(p, x, y, n, m):
  """
  polyVal2D(p, x, y, n, m)
  
  Evaluate a 2-D polynomial using Horner's method.
  
  Evaluates the 2-D polynomial `p` at the points specified by `x`
  and `y`, which must be the same dimensions. The output `f` will
  have the same dimensions as `x` and `y`. The order of `x` and `y`
  are specified by `n` and `m`, respectively.
  
  Parameters
  ----------
  p : array_like
      Polynomial coefficients in order specified by polyVal2D.html.
  x : array_like
      Values of 1st independent variable.
  y : array_like
      Values of 2nd independent variable.
  n : int
      Order of the 1st independent variable, `x`.
  m : int
      Order of the 2nd independent variable, `y`.
  
  Returns
  -------
  f : ndarray
      Values of the evaluated 2-D polynomial.
  
  See Also
  --------
  numpy.polyval : Evaluate a polynomial at specific values.
  
  Example
  --------
  >>> (5**2 * 6**2 + 2 * 5 * 6**2 + 3 * 6**2 + 4 * 5**2 * 6 +
  ...  5 * 5 * 6 + 6 * 6 + 7 * 5**2 + 8 * 5 + 9)
  2378
  >>> polyVal2D(np.array([1,2,3,4,5,6,7,8,9]),5,6,2,2)
  2378
  >>> 5 * 6 + 2 * 6 + 3 * 5 + 4
  61
  >>> polyVal2D(np.array([1,2,3,4]),5,6,1,1)
  61
  """
  # TODO: check input args
  x = np.array(x)
  y = np.array(y)
  f = p[0]
  for ni in range(n):
    f = f * x + p[1 + ni]
  for mi in range(m):
    mpp = (n + 1) * (mi + 1)
    g = p[mpp]
    for ni in range(n):
      g = g * x + p[mpp + 1 + ni]
    f = f * y + g
  return f

File: polyVal2D.m
function f = polyVal2D(p,x,y,n,m)
%POLYVAL2D Evaluate a 2-D polynomial using Horner's method.
%   F = POLYVAL2D(P,X,Y) evaluates the 2-D polynomial P at the points 
%   specified by X and Y, which must have the same dimensions. The output F
%   will have the same dimensions as X and  Y. N and M specify the order of X
%   and Y respectively. Polynomial coefficients are in the following order.
%
%   F(X,Y) = P_1 * X^N * Y^M + P_2 * X^{N-1} * Y^M + ... + P_{N+1} * Y^M + ...
%            P_{N+2} * X^N * Y^{M-1} + P_{N+3} * X^{N-1} * Y^{M-1} + ... + P_{2*(N+1)} * Y^{M-1} + ...
%            ...
%            P_{M*(N+1)+1} * X^N + P_{M*(N+1)+2} * X^{N-1} + ... + P_{(N+1)*(M+1)}
%
%
% See also: POLYFITN by John D'Errico on MathWorks MATLAB Central FEX
% http://www.mathworks.com/matlabcentral/fileexchange/34765-polyfitn
%% check input args
validateattributes(p,{'numeric'},{'2d','nonempty','real','finite'}, ...
    'polyVal2D','p',1)
validateattributes(x,{'numeric'},{'nonempty','real','finite'}, ...
    'polyVal2D','x',2)
validateattributes(y,{'numeric'},{'nonempty','real','finite'}, ...
    'polyVal2D','y',3)
assert(all(size(x)==size(y)),'polyVal2D:sizeMismatch', ...
    'X and Y must be the same size.')
% use size of p to set n & m
pdims = size(p);
if nargin<4 && all(pdims>1)
    n = pdims(1)-1;
    m = pdims(2)-1;
end
validateattributes(n,{'numeric'},{'scalar','integer','positive','<',10}, ...
    'polyVal2D','n',4)
validateattributes(m,{'numeric'},{'scalar','integer','positive','<',10}, ...
    'polyVal2D','m',5)
if all(pdims>1) 
    assert(pdims(1)==n+1,'polyVal2D:xOrderMismatch', ...
        'The number of x coefficients doesn''t match the order n.')
    assert(pdims(2)==m+1,'polyVal2D:yOrderMismatch', ...
        'The number of y coefficients doesn''t match the order m.')
end
%% evaluate polynomial P
f = p(1);
for ni = 1:n
    f = f.*x+p(1+ni);
end
for mi = 1:m
    mj = (n+1)*mi+1;
    g = p(mj);
    for ni = 1:n
        g = g.*x+p(mj+ni);
    end
    f = f.*y+g;
end
